cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+do the same, within reason, so a helper library will be made available. Originally it was thought that such a helper
+library was not needed and extensions could deal with this themselves, but it is clear that there are too many common
+flags and situations to have to rewrite logic for.
+
+A package on the existing Go CLI will be made available for use programmatically. The package is
+`github.com/temporalio/cli/cliext` and is the only package on the entire library/module that is acceptable for use
+programmatically. The existing `github.com/temporalio/cli/temporalcli` package should not be used and may be moved to
+`internal` as part of this project.
+
+The CLI module is versioned the same as the CLI binary version. The version of CLI used by users can be different than
+the version of CLI module used by extensions, but there may be issues if the CLI version used by a user is newer than
+the one used by an extension mostly due to new flags and Temporal client capabilities. Within reason, the CLI team will
+try to maintain runtime behavior compatibility with extensions using older forms of this library.
+
+Regarding API compatibility, unlike SDKs, there are no guarantees that `github.com/temporalio/cli/cliext` library API
+will remain compatible from one version to the next. CLI team will try its best to retain compatibility, or if it can't,
+will try to have clear compilation breaks and release notes. CLI binary versions are not meant to be construed as
+related to compatibility of this package which technically means a CLI patch could have a compatibility change in this
+library, though CLI team will strive to avoid that at the semver patch level (and only do it at a minor level).
+
+Documentation of this library is in the Godoc of the `github.com/temporalio/cli/cliext` package. Even release notes will
+not mention the library unless there is significant reason such as a compatibility break.
+
+The initial implementation of the library will contain the following utilities:
+
+* Structs and Cobra flag sets for root-level flags and Temporal client flags
+* Ability to create `*slog.Logger` from root-level flags
+* Ability to dial a Temporal client from Temporal client flags
+* Utility to create payloads from raw data
+
+Other items can be added as needed. Extensions can/should also use the Go SDK as needed.
+
+There is not a plan to expose a printer at this time, so extensions will have to handle the possible `output` enumerates
+of `text`, `json`, `jsonl`, and `none` themselves. This is because the current CLI printer has too many quirks and is
+not high quality enough for exposure. It is possible in the future a good printer abstraction can be exposed.
+
+It is an area of active research whether built-in commands will leverage this package or whether both will leverage
+common code independently.
+
+It is an area of active research whether the CLI's YAML-based code generation functionality will be made available to
+extensions.
+
```

Comment:
How will errors from extensions be presented? (Will there be any effort to make this have a consistent appearance, e.g. will the real CLI intercept and re-present stderr and/or exit codes from user executables?)

Reply:
There is no effort to make this any of the output consistent at this time. Extensions are free to do with stdin, stdout, stderr, and exit codes whatever they want.